python/pypto/backend/pto_backend.py

```
@@ -332,30 +332,82 @@ def _generate_arg_unpacking(func: _ir_core.Function, *, uses_spmd: bool = False)
+        return None
+
+    def _extract_from_dim_expr(
+        dim_expr: object, target_var: _ir_core.Var, tensor_name: str, dim_idx: int
+    ) -> str:
+        shape_expr = f"static_cast<int64_t>({tensor_name}_tensor->shapes[{dim_idx}])"
+        # Direct mapping: dim is exactly the dynamic var.
+        if isinstance(dim_expr, _ir_core.Var) and dim_expr.same_as(target_var):
+            return shape_expr
+        # Invert common affine form: dim = var * C (or C * var).
+        if isinstance(dim_expr, _ir_core.Mul):
+            l, r = dim_expr.left, dim_expr.right
```

Ruff (0.15.12)

[error] Ambiguous variable name: `l`

(E741)